communication/src/TlsConnectionManager.cpp

```diff
@@ -311,24 +313,36 @@ void ConnectionManager::resolve(NodeNum i) {
 }
 
 void ConnectionManager::connect(NodeNum i, const asio::ip::tcp::endpoint& endpoint) {
-  auto [it, inserted] = connecting_.emplace(i, asio::ip::tcp::socket(io_context_));
+  auto [it, inserted] =
+      connecting_.emplace(i, std::make_pair(asio::ip::tcp::socket(io_context_), asio::steady_timer(io_context_)));
   ConcordAssert(inserted);
   status_->num_connecting = connecting_.size();
-  it->second.async_connect(
+  LOG_TRACE(logger_, "connecting to node : " << i);
+  // If async_connect takes too long to finish
+  // the timeout will happen after given duration
+  connecting_.at(i).second.expires_after(CONNECT_DEADLINE);
+  it->second.first.async_connect(
       endpoint, asio::bind_executor(strand_, [this, i, endpoint](const auto& error_code) {
         if (error_code) {
           LOG_WARN(logger_, "Failed to connect to node " << i << ": " << endpoint << " : " << error_code.message());
-          connecting_.at(i).close();
-          connecting_.erase(i);
+          if (connecting_.find(i) != connecting_.end()) {
+            connecting_.at(i).first.close();
+            LOG_WARN(logger_, "socket closed ");
+            connecting_.at(i).second.cancel();
+          }
           status_->num_connecting = connecting_.size();
           return;
         }
         LOG_INFO(logger_, "Connected to node " << i << ": " << endpoint);
-        auto connected_socket = std::move(connecting_.at(i));
-        connecting_.erase(i);
+        auto connected_socket = std::move(connecting_.at(i).first);
+        // cancelling the timer
+        connecting_.at(i).second.cancel();
+
```

Comment:
erase is there in the callback fun of timer expiry so when timer cancel the callback will erase from the map

Reply:
I don't think it is the correct approach. In case the connection is successfully established, we need to cancel the timer and remove the connection from connecting_ list.